### PR TITLE
Added Kick and Chip Primitive classes

### LINKED ISSUE #87: Create Shoot Primitive class
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
We need a class in the AI for the Shoot Primitive, so the AI can use it as an output.

**Describe the solution you'd like**
Implement a Shoot Primitive class that extends from the base Primitive class. This will go in the `thunderbots/software/ai/primtiive` folder. See the `move_primitive.*` files for an example.

Make sure to document the spec of the primitive, using the information from #86. This can probably go at the top of the class (in a big javadoc style comment).

### PR DESCRIPTION
**Pull request by jessicahe21**
### Description
Added a ~~Shoot~~ Kick and Chip Primitive classes under ~~`shoot_primitive.*`~~ `kick` and `chip_primitive.*` that extends from the base Primitive class. 

Follows similar formatting as `move_primitive.*` such as the commenting style. 
(This is a follow-up PR to #121 )

### Testing Done
Unit tests are done for all KickPrimitive and ChipPrimitive class functions.
Running `catkin_make` builds the code / AI.

### Resolved Issues
resolves #87 , part of #135 

### Review Checklist
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

I am not certain if I referred to the extra bit array correctly for chipping in line 22 of `shoot_primitive.cpp` and any guidance would be greatly appreciated.
As well, please let me know if I am missing some comments for the `.cpp` and `.h` (before the class) files as I referred to `move_primitive.*` for commenting style.

### CODE REVIEW

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/shoot_primitive.h]: We had some discussion about using `power` on a recent PR: https://github.com/UBC-Thunderbots/Software/pull/108#discussion_r226637254
(Tl;DR - power can be a bit of a confusing word (ie. it's technically work per unit time))

I'm just guessing at what this is supposed to be, but maybe `kick_speed_meters_per_second` would be a bit more descriptive?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
This is looking good, just one small comment.

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: Yea, I'd hesitate a bit more here to call it `kick_speed_meters_per_second` because the units are not always meters per second (aka speed). I was ok with `kick_speed_meters_per_second` in the context of grsim because `m/s` are the units for both kicking and chipping.

For our robots, the units when referring to a kick are `m/s` for the speed of the kick. When we are chipping, the units are just `meters`, for the distance between the robot and the location of the first bounce of the ball.

The different units in different cases make it harder to name, and I'd also rather not split this into 2 variables (eg. `kick_speed_meters_per_second` and `chip_distance_meters`) because you'd effectively have a duplicate variable which also adds complexity.

I summon @EvanMorcom @henrymliu @akhilveeraghanta and @jordanzliu for their opinions! :crystal_ball:

**Review by MathewMacDougall — COMMENTED** [review]
Followed up on @garethellis0 's comment. Other than that, looks really good!

For future reference, if CI fails due to clang-format you can just make another commit that formats the code and push the new changes to update the PR. No need to open another one (I do this all the time).

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: Hmm, I agree about not splitting variables, but looking at this from somewhat of an "external" perspective, it feels like we should have two primitives for `chip`, and `kick` respectively. For that matter, what differentiates `shoot` from just a kick?

Comment (jessicahe21) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: I agree that `power` can be more descriptive since there are two different possible functionalities to it. Is there any possible way to, say, create the variable under a conditional statement (is that how they call `if` and stuff?) after determining whether it's a `chip` or a `kick`? Or would that be even more complicated than setting up two separate variables.

Comment (jessicahe21) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: Thanks all for the feedback, by the way!

Comment (akhilveeraghanta) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: This seems like a textbook example of logical cohesion. We shouldn't have a flag that determines what the `power` variable means imo, we should definitely split `shoot` into `kick` and `chip` primitives as Gareth suggested, and give them their own definitions of `power`.

Comment (henryymliu) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: I think I'll tentatively agree with splitting the primitives. Shouldn't be too much work on the high level side.

Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: > 1. Our current radio protocol only support 8 unique primitives. We should make sure we can still support all the other primitives we need if we are adding another one. This could change if/when we move to wifi or a better radio, but that won't be for at least a year.

I don't think limiting ourselves in terms of what we can do in `ai` based on the limitations of the radio protocol like this isn't a great idea. I'd be ok with encoding them into the "same" primitive (though it's still not a great solution). No chance of just changing the radio protocol I suppose?

> 2\. A bit more nitpicky, but we'll want a better way to share code on the robot firmware. Seems like the implementations for `shoot` and `chip` on the robot would be exactly the same except for the chip/kick flag

This doesn't seem like it's really related to this PR? More of a firmware implementation thing?

Comment (EvanMorcom) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: > This seems like a textbook example of logical cohesion. We shouldn't have a flag that determines what the `power` variable means imo, we should definitely split `shoot` into `kick` and `chip` primitives as Gareth suggested, and give them their own definitions of `power`.

I Agree with this.

**Review by henryymliu — CHANGES_REQUESTED** [review]
Minor nit, and adding on to the existing conversation.

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (jessicahe21) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: I have tentatively split `shoot` into separate `kick` and `chip` primitives (didn't remove the files for `shoot`, just in case). It's true that the implementations are basically exactly the same except for the content of `power`. 
Just out of curiosity, what are the downsides of not having a one-to-one mapping between a high-level primitive and a firmware primitive?

**Comment by jessicahe21** [comment]
Added `chip_primitive.*` and `kick_primitive.*`, implementations are the basically the same as `shoot` except the extra bit array is not being used anymore. 
`power` is split into `kick_speed_meters_per_second` and `chip_distance_meters` for the two different primitives.

Also fixed minor typos in both `shoot_primitive.h` and `move_primitive.h` as Akhil pointed out + added more comments before primitive classes.

P.S. I keep on forgetting to format the files before pushing! :sob:

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (jessicahe21) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: Thanks for pointing it out, I was wondering if the existing comments were sufficient.

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: I might be wrong here (@henrymliu @MathewMacDougall please correct me if I am), but I think it just makes it a bit easier for people to reason about things. For example, if I send a high-level `kick` primitive to the robot and the robot does the wrong thing, the first place I might think to go look is in the `kick` primitive on the robot side. When there's not a one-to-one mapping however, I might need to trace the path that the `kick` primitive is taking, see that it gets mapped to a `shoot` primitive, and _then_ could figure out where in the robot firmware I could look for the potential error.

**Comment by garethellis0** [comment]
Looking good! Some points:

1. Don't worry about formatting every commit, it's totally fine to just push another commit like "Fixing Formatting" (and indeed it's pretty common, there are _lots_ of commits like that on many PR's :stuck_out_tongue:)
2. We can probably remove all the `shoot_*` files now. (In general, don't be afraid of deleting code after you've committed it. Once you've committed it to git, it's never really lost, you can always look backwards in the git history and grab it with `git checkout SOME_OLD_GIT_HASH_HERE FILE_YOU_WANT_HERE`)

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: What @garethellis0 said is correct. When things have a 1 to 1 mapping it's simpler and easier to reason about the flow of control and what changes affect what. I consider mapping two high-level primitives to a single firmware primitive a "trick", and the more tricks code has the harder it is to fully understand (quickly). However I do think this is a good solution.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
This all looks good to me :+1:. I think we can just remove the old `shoot` files and this should be good to go.

File: src/thunderbots/software/ai/primitive/shoot_primitive.h
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/shoot_primitive.h]: ^ I'm also super happy to see this kind of discussion happening with so many people :grin:. It's great to see everyone working together to learn and get to the best solution.

**Review by garethellis0 — DISMISSED** [review]

**Comment by garethellis0** [comment]
LGTM :+1: :tada:

**Review by MathewMacDougall — DISMISSED** [review]
LGTM! :+1: Just looks like you'll need to update your branch. Ping us if you need re-approval.

**Comment by jessicahe21** [comment]
Looks like I messed up my branch by doing a force pull two days ago 😱 Fingers crossed that it is working.

Though now the commit messages are a little vague and I was trying to fix it with [this guide](https://help.github.com/articles/changing-a-commit-message/), but it feels like I'm further down the rabbit hole... After updating the message with `git commit --amend` I did a force push. But now the files are gone again I think.

Any help or advices would be greatly appreciated 😭 @garethellis0 @MathewMacDougall

**Comment by MathewMacDougall** [comment]
Sounds like @garethellis0's got this, but feel free to loop me in too if necessary.

**Comment by MathewMacDougall** [comment]
@jessicahe21 it would be great if you could also add some unit tests for this class as well. That will give us test coverage right away, and is a good introduction to testing in our codebase.

You can use #131 as an example. Sorry for the last minute change.

**Review by garethellis0 — COMMENTED** [review]
LGTM, excepting the changes @MathewMacDougall mentioned above

**Comment by jessicahe21** [comment]
Unit tests are now added for all class functions of Kick and Chip Primitives! Also added 3 get functions in each primitive class like Evan did.
@MathewMacDougall Thanks for letting me know!

**Comment by garethellis0** [comment]
Looks like this is failing to compile in CI because of some wierd gtest thing. @jessicahe21 - do the tests compile and run on your local machine?

**Comment by jessicahe21** [comment]
Didn't realize `catkin_make` wasn't the right one, just ran `catkin_make run_tests` and fixed some typo problems. @garethellis0 

Though it still won't compile and returns the error
```
CMakeFiles/Makefile2:89: recipe for target 'CMakeFiles/tests.dir/rule' failed
make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
```
I think perhaps `../test/../primitive.cpp` is screwing me up. I see that MovePrimitive is used as the unit test for `primitive.cpp`, would it be necessary to include Chip and Kick in there as well?

**Comment by jessicahe21** [comment]
*Update:*
Tests now compiles successfully after Haomiao's suggestions to:
1. Add other primitives' `.cpp` files under each other in `CMakeLists.txt` 
2. Extend the if-statement in `/ai/../primitive.cpp`

**Comment by garethellis0** [comment]
LGTM :+1: :tada: 

- paging @henrymliu, as I believe his current review will block merging until approved
- @jessicahe21, can you change `partly resolves #135` to `part of #135` in your pull request description before merging this? Otherwise I think github will automatically close #135 when we merge this.

File: src/thunderbots/software/ai/primitive/move_primitive.h
Comment (henryymliu) [inline src/thunderbots/software/ai/primitive/move_primitive.h:18]: Thanks for adding this :D

**Review by henryymliu — DISMISSED** [review]
LGTM, awesome work!

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :tada: :+1:

**Comment by MathewMacDougall** [comment]
@jessicahe21 just update your branch and we should be able to merge and wrap this up.

**Comment by MathewMacDougall** [comment]
@jessicahe21 looks like there are some build failures relating to the new primitives.

**Comment by jessicahe21** [comment]
@MathewMacDougall  Yes, I’m trying to identify the issue right now. My assumption is that the if-statement for the new primitives weren’t extended in ‘ai/../primitive.cpp’. Found minor typos for catch primitive, made a comment on the PR.

**Comment by MathewMacDougall** [comment]
@jessicahe21 any luck getting the build failures resolved? Looks like your branch will need to be updated from `master` too

**Comment by jessicahe21** [comment]
@MathewMacDougall Unfortunately no, I am still having troubles with build failures after pulling the new merges. Attempted to tinker `catch_primitive` but I think I'm further down the rabbit hole. I'm a little lost as in why the tests are failing.

**Comment by MathewMacDougall** [comment]
To clarify, is the code failing to _build_, or are the unit tests not passing?
Based on the last thing I can see from TravisCI, there's an undefined reference to `PRIMITIVE_NAME` in the `ChipPrimitive` class, but it is defined in the `ChipPrimitive` class. I'll see if I can take a closer look tonight.

**Comment by garethellis0** [comment]
@jessicahe21 - we should probably deal with this via slack. I'm around all week, so feel free to DM me (or #software, or a group with a few people in it, whatever works) with exactly what issue you're dealing with and we'll get this sorted right quick! :grin:

**Review by garethellis0 — APPROVED** [review]
Tests LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1: Nice work!

**Pull request merged**
